environs/imagemetadata/simplestreams.go

```diff
@@ -189,17 +190,26 @@ func (ic *ImageConstraint) IndexIds() []string {
 
 // ProductIds generates a string array representing product ids formed similarly to an ISCSI qualified name (IQN).
 func (ic *ImageConstraint) ProductIds() ([]string, error) {
-	stream := idStream(ic.Stream)
+	streams := ic.Streams
+	if len(streams) == 0 {
+		// If there are no streams specified use the default ""
+		streams = []string{""}
+	}
 	nrArches := len(ic.Arches)
 	nrSeries := len(ic.Series)
-	ids := make([]string, nrArches*nrSeries)
-	for i, arch := range ic.Arches {
-		for j, ser := range ic.Series {
-			version, err := series.SeriesVersion(ser)
-			if err != nil {
-				return nil, err
+	nrStreams := len(streams)
+
+	ids := make([]string, 0, nrArches*nrSeries*nrStreams)
+	for _, stream := range streams {
+		stream = idStream(stream)
+		for _, arch := range ic.Arches {
```

Comment:
this could be pulled out to a separate func; inline would do

Reply:
Do you mean pull the arch and ser loops into a local func? Why do that? (I guess it would reduce rightward drift?)